Mainnet/STOContract/STOContract/STOContract.cs

```diff
@@ -0,0 +1,1361 @@
+        private set => this.PersistentState.SetAddress(nameof(MapperAddress), value);
+    }
+
+    public ulong TokenBalance
+    {
+        get => this.PersistentState.GetUInt64(nameof(TokenBalance));
+        private set => this.PersistentState.SetUInt64(nameof(TokenBalance), value);
+    }
+
+    public bool IsNonFungibleToken
+    {
+        get => this.PersistentState.GetBool(nameof(IsNonFungibleToken));
+        private set => this.PersistentState.SetBool(nameof(IsNonFungibleToken), value);
+    }
+
+    public bool SaleOpen => EndBlock >= this.Block.Number && TokenBalance > 0;
+
+    public Address Owner
+    {
+        get => this.PersistentState.GetAddress(nameof(Owner));
+        private set => this.PersistentState.SetAddress(nameof(Owner), value);
+    }
+
+    public SalePeriod[] SalePeriods
+    {
+        get => PersistentState.GetArray<SalePeriod>(nameof(SalePeriods));
+        private set => PersistentState.SetArray(nameof(SalePeriods), value);
+    }
+
+    public STOContract(ISmartContractState smartContractState,
+                       Address owner,
+                       uint tokenType,
+                       ulong totalSupply,
+                       string name,
+                       string symbol,
+                       Address kycAddress,
+                       Address mapperAddress,
+                       byte[] salePeriods) : base(smartContractState)
+    {
+        Assert(tokenType < 3, $"The {nameof(tokenType)} parameter can be between 0 and 2");
+
+        Assert(PersistentState.IsContract(kycAddress), "The kycAdress is not a contract adress");
+
+        var periods = Serializer.ToArray<SalePeriodInput>(salePeriods);
+
+        ValidatePeriods(periods);
+        var tokenTypeEnum = (TokenType)tokenType;
+        var result = CreateTokenContract(tokenTypeEnum, totalSupply, name, symbol);
+
+        Assert(result.Success, "Creating token contract failed.");
+
+        Log(new STOSetupLog { TokenAddress = result.NewContractAddress });
+
+        KYCAddress = kycAddress;
+        MapperAddress = mapperAddress;
+        TokenAddress = result.NewContractAddress;
+        IsNonFungibleToken = tokenTypeEnum == TokenType.NonFungibleToken;
+        TokenBalance = IsNonFungibleToken ? ulong.MaxValue : totalSupply;
+        Owner = owner;
+        SetPeriods(periods);
+    }
+
+    private ICreateResult CreateTokenContract(TokenType tokenType, ulong totalSupply, string name, string symbol)
+    {
+        switch (tokenType)
+        {
+            case TokenType.StandardToken: return Create<StandardToken>(parameters: new object[] { totalSupply, name, symbol });
+            case TokenType.DividendToken: return Create<DividendToken>(parameters: new object[] { totalSupply, name, symbol });
+            default: return Create<NonFungibleToken>(parameters: new object[] { name, symbol });
+        }
+    }
+    public override void Receive() => Invest();
+
+    public bool Invest()
+    {
```

Comment:
I think you need to call `base.Receive()` here too

Reply:
It is an empty method so no needed

https://github.com/stratisproject/Stratis.SmartContracts/blob/36c37e259e4ab72c514f9c0553fba13e6a62da4d/Stratis.SmartContracts/SmartContract.cs#L123

Reply:
hmm maybe Cirrus 1.3 is using older versions of this. If I did not include base.Receive, tokens sent directly to the contract failed weren't actually received IIRC and the transaction failed. Might be worth testing, don't call the contract, just send CRS to it. I'll double check this on my end too